# Tensorflow Learning
## www.elecwizard.ir
### chapter 2
#### Arrays: Working with Images

In this tutorial, we are going to work with an image, in order to visualise changes to an array. Arrays are powerful structures, as we saw briefly in the previous tutorial. Generating interesting arrays can be difficult, but images provide a great option.

To load the image, we use matplotlib’s image module:

In [1]:
import matplotlib.image as mpimg
import os
from scipy import misc
import tensorflow as tf
# First, load the image

image = misc.imread('/home/reza/DeepLearning/Tensorflow/MarshOrchid.jpg')
# Print out its shape
print(image.shape)

(5528, 3685, 3)



The above code reads in the image as a NumPy array, and prints out the size. Note that the filename needs to be a full path (absolute or relative) to the downloaded image file.

You’ll see the output, which is (5528, 3685, 3). This means the image is 5528 pixels high, 3685 pixels wide, and 3 colors “deep”.

You can view the current image using pyplot, like so:



In [ ]:
import matplotlib.pyplot as plt

# This is a bit of magic to make matplotlib figures appear inline in the
# notebook rather than in a new window.
%matplotlib inline
plt.rcParams['figure.figsize'] = (10.0, 8.0) # set default size of plots
plt.rcParams['image.interpolation'] = 'nearest'
plt.rcParams['image.cmap'] = 'gray'

# Some more magic so that the notebook will reload external python modules;
# see http://stackoverflow.com/questions/1907993/autoreload-of-modules-in-ipython
%load_ext autoreload
%autoreload 2
plt.imshow(image)
plt.show()

### Geometric Manipulations

The first transformation we will perform is a transpose, turning the image 90 degrees counter-clockwise. The full program is below, most of which you have seen.

In [ ]:
# create a tensorflow variable
x=tf.Variable(image,name='x')
init=tf.global_variables_initializer()
with tf.Session() as sess:
    sess.run(init)
    x=tf.transpose(x,perm=[1,0,2])
    result=sess.run(x)
plt.imshow(result)
plt.show()

The new bit is this line:

    x = tf.transpose(x, perm=[1, 0, 2])

This line uses TensorFlow’s transpose method, swapping the axes 0 and 1 around using the perm parameter (axis 2 stays where it is).



The next manipulation we will do is a flip (left-right), swapping the pixels from one side to another.
For this function, it can be best thought of as:

    1-Iterate through the array according to batch_dim. Setting batch_dim=0 means we go through the rows (top to bottom).
    2-For each item in the iteration
    Slice a second dimension, denoted by seq_dim. Setting seq_dim=1 means we go through the columns (left to right).
    The slice for the nth item in the iteration is denoted by the nth item in seq_lengths

Lets see it in action:

In [ ]:
# create a tensorflow variable
y=tf.Variable(image,name='y')
height, width, depth = image.shape
init=tf.global_variables_initializer()
with tf.Session() as sess:
    sess.run(init)
    y= tf.reverse_sequence(y, [width] * height, 1, batch_dim=0)
    result=sess.run(y)
plt.imshow(result)
plt.show()

The new bit is this line:

     x = tf.reverse_sequence(x, np.ones((height,)) * width, 1, batch_dim=0)

It iterates over the image top to bottom (along its height), and slices left to right (along its width). From here, it then takes a slice of size width, where width is the width of the image.